### PR TITLE
Add "--attach-dependencies" to command "up" for attaching to dependencies

### PR DESCRIPTION
**Pull request by benthorner**
When using the 'up' command, only services listed as arguments are
attached to, which can be very different to the 'no argument' case
if a service has many and deep dependencies:

   - It's not clear when dependencies have failed to start. Have to run
'compose ps' separately to find out.
   - It's not clear when dependencies are erroring. Have to run 'compose
logs' separately to find out.

With a simple setup, it's possible to work around theses issue by
using the 'up' command without arguments. But when there are lots of
'top-level' services, with common dependencies, in a single config,
using 'up' without arguments isn't practical due to resource limits
and the sheer volume of output from other services.

This introduces a new '--attach-dependencies' flag to optionally attach
dependent containers as part of the 'up' command. This makes their logs
visible in the output, alongside the listed services. It also means we
benefit from the '--abort-on-container-exit' behaviour when dependencies
fail to start, giving more visibility of the failure.

Resolves https://github.com/docker/compose/issues/4611

Extends https://github.com/docker/compose/pull/2130 to treat `up <service>`, which implicitly includes dependent 
containers, the same way as `up`, which refers to all services/containers.

### CODE REVIEW

**Comment by benthorner** [comment]
Just wondering if I need to do something to get this reviewed/merged?

File: compose/cli/main.py [hunk 1]
```diff
@@ -1065,9 +1066,11 @@ def up(self, options):
         remove_orphans = options['--remove-orphans']
         detached = options.get('--detach')
         no_start = options.get('--no-start')
+        attach_dependencies = options.get('--attach-dependencies')
 
         if detached and (cascade_stop or exit_value_from):
```
Comment (ulyssessouza) [inline compose/cli/main.py]: Here, shouldn't you test with `attach_dependencies` in the `or` section?

File: compose/cli/main.py [hunk 2]
```diff
@@ -1065,9 +1066,11 @@ def up(self, options):
...
-            raise UserError("--abort-on-container-exit and -d cannot be combined.")
+            raise UserError(
+                "-d cannot be combined with --abort-on-container-exit or -attach-dependencies.")
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
                "-d cannot be combined with --abort-on-container-exit or --attach-dependencies.")
```

File: compose/cli/main.py [hunk 3]
```diff
@@ -1098,6 +1101,7 @@ def up(rebuild):
                     always_recreate_deps=always_recreate_deps,
                     reset_container_image=rebuild,
                     renew_anonymous_volumes=options.get('--renew-anon-volumes'),
+
```
Comment (ulyssessouza) [inline compose/cli/main.py]: nit: Remove new line

File: compose/cli/main.py (hunk 1 above)
Comment (benthorner) [reply compose/cli/main.py]: 👍 - fixed. I've also added it in a subsequent section, as it's not compatible with `--no-start`.

File: compose/cli/main.py (hunk 2 above)
Comment (benthorner) [reply compose/cli/main.py]: 👍 - updated.

File: compose/cli/main.py [hunk 4]
```diff
@@ -1392,10 +1398,13 @@ def log_printer_from_project(
         log_args=log_args)
 
 
-def filter_containers_to_service_names(containers, service_names):
+def filter_attached_containers(containers, service_names, attach_dependencies):
     if not service_names:
         return containers
 
+    if attach_dependencies:
```
Comment (ulyssessouza) [inline compose/cli/main.py]: Since the content of this `if` is equal to the preceding one, please, just add an `or` to the first one

File: compose/cli/main.py [hunk 5]
```diff
@@ -1392,10 +1398,13 @@ def log_printer_from_project(
         log_args=log_args)
 
 
-def filter_containers_to_service_names(containers, service_names):
+def filter_attached_containers(containers, service_names, attach_dependencies):
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
def filter_attached_containers(containers, service_names, attach_dependencies=False):
```
By using a default value we can keep the same usage.
With this you can simplify the calls in the tests that are not related to the PR.

Comment (benthorner) [reply compose/cli/main.py]: Good suggestion - updated 👍 .

File: compose/cli/main.py (hunk 4 above)
Comment (benthorner) [reply compose/cli/main.py]: Sure - I've done that.

**Comment by ulyssessouza** [comment]
@benthorner  You also need to add the new option to the [shell auto completion scripts](https://github.com/docker/compose/tree/master/contrib/completion). After that I think it's complete

**Comment by benthorner** [comment]
@ulyssessouza thanks for pointer - I completely missed that before. Should be good for another look now 👀 .

**Comment by ulyssessouza** [comment]
Thank you for the contribution @benthorner !

**Comment by thaJeztah** [comment]
LOL, was just about to comment: wondering why this should not be the default?

```bash
# starts all containers and attaches to them
docker-compose up

# start only service-a and service-b, including their dependencies, and attaches to them
docker-compose up <service-a> <service-b>
```

I see the original issue mentions some people only being interested in the output of the service, not their dependencies (which is inconsistent with `docker-compose up` (all services)).

For those, it would be possible to start the services _detached_

```bash
docker-compose up -d <service-a> <service-b>
```

And attach to the logs of those services that they want to view the logs for

```bash
docker-compose logs -f <service-a> <service-b>
```

**Comment by benthorner** [comment]
@thaJeztah I think the previous behaviour was consistent: `docker-compose up [services...]` would only attach to the listed services; the only special is case is with no arguments, which is interpreted as "all services". So the new behaviour is different (and therefore not the default), since in the first case it will cause non-listed service output to be shown.

> Extends #2130 to treat up <service>, which implicitly includes dependent
containers, the same way as up, which refers to all services/containers.

I notice I'm sort of contradicting myself 😆.
